### ISSUE TITLE
Reference Editor onCreate

### ISSUE THREAD

**Issue by BraunreutherA** [contentful/field-editors#586]
Hi there, 

the onCreate prop passed to the `LinkActions` in the Reference Editor are forcing the developer to a specific handling of the editor as it creates a new entry in the background and opens the drawer. Would it be possible to also pass the onCreate Callback of the  https://github.com/contentful/field-editors/blob/c6db8bf095f1f1cf2b3de610f34bbe483098ca04/packages/reference/src/common/MultipleReferenceEditor.tsx down to the `renderCustomComponents` so the dev can build custom actions like creating a copy of the active entry and open this?

**Comment by anho** [comment]
Hey,

I assume you are talking about `renderCustomActions` in both the `SingleReferenceEditor` and `MultipleReferenceEditor`. 

We currently don't have a use case for that internally, so there is little chance we will do that soon. But let me assure you, that if you feel a strong need for your particular use-case open a PR and we will happily review it.

Cheers

**Comment by BraunreutherA** [comment]
Hi @anho, 

I've created a PR #604 which passes the native `onCreate` and `onLink` through to the  renderCustomActions method. This way t's possible to build completely custom actions. 
I've named the functions `onEntryCreated` and `onEntryLinked` - hope it fits your style. 

Best 
Alex

**Comment by djagya** [comment]
@BraunreutherA Hi,

Reference editor also accepts [`onAction` property](https://github.com/contentful/field-editors/blob/master/packages/reference/src/common/ReferenceEditor.tsx#L23) that [gets called](https://github.com/contentful/field-editors/blob/master/packages/reference/src/components/LinkActions/LinkEntityActions.tsx#L36-L43) when an entry is created/linked. 

Can it be used to do something similar? Like in your case, if I understood it correctly: an entry gets created when clicking "create" in the editor, then you want to duplicate it and open it instead? Then in `onAction` you'll get `create_and_link` event and you can then duplicate an entry there and open it.

**Comment by BraunreutherA** [comment]
Hi @djagya,

the `onAction` approach is the way we do it right now. Sadly it has some side effects which are not very nice in this case: 
- The Drawer opens immediately before the content is opened to the new entry. That way a flicker effect exists as it shows a loading screen, then shows the empty entry and the after a second or two shows the copied content. We'd like to open the drawer after the entry was copied. 
- We're relying of the relation data model to show specific fields or hide them in the copied component. This is done by a query to show if other files have a relation towards the opened entry. As the opening is done async even before the relation is built in the database we have to poll it, which is not good for the customers query count aswell it also creates unenecessary flicker effects. 

With a synchronous way we can avoid these problems.

**Comment by djagya** [comment]
@BraunreutherA sorry for the delay. I have a couple more questions, just to clarify that I understood it correctly.

If I understand correctly, you want to copy an entry and then open it. The problem with passing `onEntryCreated` as in you PR is that it will be called only after [this line creating and entry and opening it in a slide-in editor](https://github.com/contentful/field-editors/blob/master/packages/reference/src/components/LinkActions/LinkEntityActions.tsx#L31) is executed.

Why not override the `onCreate(contentTypeId?: string, index?: number)` property instead? Inside you could duplicate an entry, open it etc. But I guess the problem is that it's still required [to call `props.onCreate`](https://github.com/contentful/field-editors/blob/master/packages/reference/src/components/LinkActions/LinkEntityActions.tsx#L35) to notify the editor about a new entry.

So for you case, instead of [creating and opening an entry](https://github.com/contentful/field-editors/blob/master/packages/reference/src/components/LinkActions/LinkEntityActions.tsx#L31-L34) you'd be doing you custom stuff in a custom `onCreate` callback and then calling `onEntityCreated` that [would contain this](https://github.com/contentful/field-editors/blob/master/packages/reference/src/components/LinkActions/LinkEntityActions.tsx#L35-L43), right?

**Comment by BraunreutherA** [comment]
@djagya That's exactly the plan. I'd have a button with a onClick listener which creates an entry directly through the app sdk, and sets the fields to the values of the copied entry. 
When this is done I would call sdk.navigator.openEntry and the onEntryCreated to finally add it to the list of references in the Reference Editor field.

> Why not override the onCreate(contentTypeId?: string, index?: number) property instead? Inside you could duplicate an entry, open it etc. But I guess the problem is that it's still required to call props.onCreate to notify the editor about a new entry.

The onCreate here needs to be called so the entry gets added as a reference. But onCreate as implemented right now does the whole creation and opening not just the adding of the link. 

What I would need is a callback to tell the ReferenceEditor that I've created a entry and it should add it to the list of reference links.

**Comment by djagya** [comment]
@BraunreutherA ok, got it, thanks for the explanation! In that case I'll make a new PR with a few more additions, so `onAction` will also be triggered when `onCreated` is called.

**Comment by djagya** [comment]
@BraunreutherA thanks for your contribution. I covered the described use-case in my PR and added some examples, so I'll close your PR if you don't mind.

**Comment by BraunreutherA** [comment]
@djagya That's awesome - thank you so much! I also like the implementation with the complete action setup :) 
Thank you for your help!

**Comment by djagya** [comment]
Hi @BraunreutherA, the PR is merged, the new props are available in `@contentful/field-editor-reference@2.10.0`. The web app now also uses the latest version. Please let me know if it allowed you to better solve the problem you described!

For now I'm closing the issue, thanks for patience :)

**Comment by BraunreutherA** [comment]
Hi @djagya, I've implemented our feature by using the new callbacks and it worked like a charm. Thanks a lot! :)
